library/src/cache/patch_manager.rs

```diff
@@ -244,28 +244,30 @@ impl PatchManager {
             .with_context(|| format!("Failed to delete patch dir {}", &patch_dir.display()))
     }
 
-    /// Attempts to use the last successfully booted patch as the next boot patch. If the last successfully
-    /// booted patch is not bootable or has the same number as the patch we're falling back from, we clear it.
-    fn try_fall_back_to_last_booted_patch(&mut self) {
+    /// Deletes artifacts for the provided bad_patch_number and attempts to set the next_boot_patch to the last
+    /// successfully booted patch. If the last successfully booted patch is not bootable or has the same number
+    /// as the patch we're falling back from, we clear it as well.
+    fn try_fall_back_from_patch(&mut self, bad_patch_number: usize) {
+        // No need to log failure – delete_patch_artifacts logs for us.
+        let _ = self.delete_patch_artifacts(bad_patch_number);
+
         if let Some(next_boot_patch) = self.patches_state.next_boot_patch {
-            // If we have a next_boot_patch that we're falling back from, delete its artifacts.
-            let _ = self.delete_patch_artifacts(next_boot_patch.number);
-            self.patches_state.next_boot_patch = None;
-
-            if let Some(last_boot_patch) = self.patches_state.last_booted_patch {
-                if last_boot_patch.number == next_boot_patch.number {
-                    // If the last booted patch is the same as the next boot patch, clear it.
-                    self.patches_state.last_booted_patch = None;
-                }
+            // If our next boot patch is bad_patch_number, clear it.
+            if next_boot_patch.number == bad_patch_number {
+                self.patches_state.next_boot_patch = None;
             }
         }
 
+        // If we think we can still boot from the last booted patch, set it as the next_boot_patch.
+        // If something happened to render the last boot patch unbootable, clear it and delete its artifacts.
         if let Some(last_boot_patch) = self.patches_state.last_booted_patch {
-            if self.validate_patch_is_bootable(&last_boot_patch).is_ok() {
-                // If we think we can still boot from the last booted patch, set it as the next_boot_patch.
+            if last_boot_patch.number != bad_patch_number
```

Comment:
I'm still not sure why bad_patch_number ever matters to a last_boot_patch?

If we're trying to catch the case where:
- success boot patch 1
- install patch 2
- fail to boot patch 2
- fail to boot patch 1
then lets explicitly test that.

Previously I was under the assumption that:

- success boot patch 1
- install patch 2
- fail boot patch 2
- fail boot patch 1 (ignore failure)
- keep trying patch 1 since we succeeded once.

I'm fine with us having a "we will roll back until we successfully boot" behavior,

What about?
- success boot patch 1
- fail boot patch 1
Do we roll back in that case?  Previously I was under the impression we should not.

Regardless, I'd encourage testing all of the above.

Reply:
The first case you mention isn't exactly right, as we're checking here whether we should even try to boot patch 1 as a fallback. What we currently try to handle now is:

Case A:
1. Patch 1 boot success
2. Patch 2 boot failure
3. Patch 1 still bootable?
4. If so, try boot patch 1
5. If not, fall back to release

> Do we roll back in that case? Previously I was under the impression we should not.

Yes, we do and should. Otherwise we'd end up repeatedly trying to boot patch 1 and effectively brick the app.

Calling this scenario "Case B":
- success boot patch 1
- fail boot patch 1

We test both cases between the `fall_back_tests` and the `next_boot_patch_tests` (because `next_boot_patch` is the public API that exercises this code):

- Case A:
  - `falls_back_to_last_booted_patch_if_still_bootable`
  - `does_not_fall_back_to_last_booted_patch_if_corrupted`
  - `clears_next_and_last_patches_if_both_fail_validation`
- Case B:
  - `returns_none_if_next_boot_patch_is_not_bootable`, 
  - `clears_current_and_next_on_boot_failure_if_they_are_the_same` 

Let me know if you feel like something is missing here. AFAICT, we test all fallback behavior cases and verify that the correct artifacts are left on disk afterwards in the tests as they currently stand.

Reply:
Thank you for the thorough response.  This is definitely fodder for some later writings on safety & security.